File: src/util/entities/AutomodRule.ts
/*
	Spacebar: A FOSS re-implementation and extension of the Discord.com backend.
	Copyright (C) 2024 Spacebar and Spacebar Contributors

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU Affero General Public License as published
	by the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU Affero General Public License for more details.

	You should have received a copy of the GNU Affero General Public License
	along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

import { BaseClass } from "./BaseClass";
import { Entity, JoinColumn, ManyToOne, Column } from "typeorm";
import { User } from "./User";
import { AutomodAction, AutomodRuleActionType, AutomodRuleEventType, AutomodRuleTriggerMetadata, AutomodRuleTriggerType } from "@spacebar/schemas";

@Entity({
	name: "automod_rules",
})
export class AutomodRule extends BaseClass {
	@JoinColumn({ name: "creator_id" })
	@ManyToOne(() => User, { onDelete: "CASCADE" })
	creator: User;

	@Column()
	enabled: boolean;

	@Column()
	event_type: AutomodRuleEventType;

	@Column({ type: "simple-array" })
	exempt_channels: string[];

	@Column({ type: "simple-array" })
	exempt_roles: string[];

	@Column()
	guild_id: string;

	@Column()
	name: string;

	@Column()
	position: number;

	@Column()
	trigger_type: AutomodRuleTriggerType;

	@Column({
		type: "simple-json",
		nullable: true,
	})
	trigger_metadata?: // this is null for "Block suspected spam content"
	| AutomodRuleTriggerMetadata;

	@Column({
		type: "simple-json",
	})
	actions: AutomodAction[];
}

File: extra/admin-api/Spacebar.Db/Models/AutomodRule.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;

namespace Spacebar.Db.Models;

[Table("automod_rules")]
public partial class AutomodRule
{
    [Key]
    [Column("id", TypeName = "character varying")]
    public string Id { get; set; } = null!;

    [Column("enabled")]
    public bool Enabled { get; set; }

    [Column("event_type")]
    public int EventType { get; set; }

    [Column("exempt_channels")]
    public string ExemptChannels { get; set; } = null!;

    [Column("exempt_roles")]
    public string ExemptRoles { get; set; } = null!;

    [Column("guild_id", TypeName = "character varying")]
    public string GuildId { get; set; } = null!;

    [Column("name", TypeName = "character varying")]
    public string Name { get; set; } = null!;

    [Column("position")]
    public int Position { get; set; }

    [Column("trigger_type")]
    public int TriggerType { get; set; }

    [Column("trigger_metadata")]
    public string? TriggerMetadata { get; set; }

    [Column("actions")]
    public string Actions { get; set; } = null!;

    [Column("creator_id", TypeName = "character varying")]
    public string? CreatorId { get; set; }

    [ForeignKey("CreatorId")]
    [InverseProperty("AutomodRules")]
    public virtual User? Creator { get; set; }
}
